### ISSUE TITLE
Model from/to pointer

### ISSUE THREAD

**Issue by AntoinePrv** [SCIP-Interfaces/PySCIPOpt#330]
Hello,

We're developing a [library](github.com/ds4dm/ecole) with Python bindings using Scip in the background and we would like our user to be able to leverage PyScipOpt for their own customization.

The easiest for us would be to exchange a `SCIP*` through Python, using [`ctypes.c_void_p`](https://docs.python.org/3.8/library/ctypes.html#ctypes.c_void_p).
For PyScipOpt, I would see this as adding two methods:

```python
class Model:
    ...

    @classmethod
    def from_ptr(cls, ptr: ctypes.c_void_p, take_ownership: bool) -> cls: ...

    def to_ptr(self, give_ownership:bool) -> ctypes.c_void_p: ...
```
Where ownership is about de-allocating or not.
I'd happily send a PR if you're interested 🙂

**Comment by fserra** [comment]
I have a question. Why would you like to have access to the pointer? The pointer is currently hidden in Model, so a solution would also be to be able to create a model from another model. Would this be enough?

**Comment by AntoinePrv** [comment]
Yes, the point is to create a `scip.Model` from a `ecole.Model` and vice versa.

The idea is to do it at run-time to make the dependency optional (I wouldn't even know where to start to expose a cython API to another extension at compile time).

Even if we put some sugar around it, like `from/to_pyscipopt` which we're doing, we still need to bring the pointer value to Python.
On your end, you probably don't want to add a new method for every library that wants to.

We could make the methods `_from_ptr` or `__from_ptr` to indicate that they are private.

**Comment by fserra** [comment]
what is an ecole.model? So you are saying that somehow you are creating an scip pointer without a pyscipopt.Model and you want to use that pointer to create a model? Or are all the scip pointers inside a pyscipopt.Model?

**Comment by AntoinePrv** [comment]
> you are creating an scip pointer without a pyscipopt.Model

Yes, we have a C++ library that is independent from Python, and create Scip pointer itself. `ecole.Model` are a bindings to that library.

> you want to use that pointer to create a model?

To create a `scip.Model`, yes.
And we would also like to transfer make it possible to transfer ownership of `scip.Model` to our classes.

**Comment by fserra** [comment]
ah ok I see, then everything is alright

**Comment by fserra** [comment]
ok so just for future reference, the idea here is to be able to create a SCIP pointer outside pyscipopt and then attach it to a pyscipopt model.

**Comment by AntoinePrv** [comment]
Hi,

I have successfully use this to share back and forth a `SCIP*` between PyScipOpt and our other extension. 🎉 

After a bit of digging, I learned that the _correct_  way of sharing pointers between Python extensions is to use a [PyCapsule](https://docs.python.org/3/c-api/capsule.html).
They basically hold a `void*`, which is [more appropriate](https://stackoverflow.com/a/40941933/5862073) than an int (although I don't think we'll run into issues with the current implementation).
I could send a PR-1 for this if you are interested. It should be a rather small change.

Furthermore, `PyCapsule` can also hold a destructor that will be called by the garbage collector.
Essentially, the destructor + Python reference counting would act as a `std::shared_ptr`.
If we use the destructor to call `SCIPfree`, and access the `SCIP*` through the PyCapsule then PyScipOpt could easily share ownership of the `SCIP*` with whoever wants them without worrying about who is going to deallocate it.
I've played a bit with it and could send a PR-2 for it as well if you are interested (longer term).
